src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2794 @@
+        }
+    }
+
+    ID3D12Fence *fence = NULL;
+    {
+        HRESULT result = (*device)->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence));
+        if (D3DError(result, fence, NULL, "Unable to create the Direct3D 12 fence for command queue")) {
+            return NULL;
+        }
+    }
+
+    queue_fence = fence;
+    __atomic_store_n(&queue_last_signal, 0, __ATOMIC_SEQ_CST);
+
+    return reinterpret_cast<d3d12_command_queue*>(commandQueue);
+}
+
+template<D3D12_COMMAND_LIST_TYPE Type>
+static d3d12_command_allocator *new_command_allocator(d3d12_device *device) {
+    TRACELOG;
+    halide_assert(user_context, device);
+    ID3D12CommandAllocator *commandAllocator = NULL;
+    HRESULT result = (*device)->CreateCommandAllocator(Type, IID_PPV_ARGS(&commandAllocator));
+    if (D3DError(result, commandAllocator, NULL, "Unable to create the Direct3D 12 command allocator")) {
+        return NULL;
+    }
+    return reinterpret_cast<d3d12_command_allocator*>(commandAllocator);
+}
+
+WEAK void add_command_list_completed_handler(d3d12_command_list *cmdList, struct command_list_completed_handler_block_literal *handler) {
+    TRACELOG;
+    TRACEPRINT("... ignoring ...\n");
+    // there's no equivalent to Metal's 'addCompletedHandler' in D3D12...
+    // the only sensible way to emulate it would be to associate an event object
+    // with a fence via ID3D12Fence::SetEventOnCompletion() and have a background
+    // thread wait on that event and invoke the callback
+}
+
+template<D3D12_COMMAND_LIST_TYPE Type>
+static d3d12_command_list *new_command_list(d3d12_device *device, d3d12_command_allocator *allocator) {
+    TRACELOG;
+    ID3D12GraphicsCommandList *commandList = NULL;
+    UINT nodeMask = 0;
+    ID3D12CommandAllocator *pCommandAllocator = (*allocator);
+    ID3D12PipelineState *pInitialState = NULL;
+    HRESULT result = (*device)->CreateCommandList(nodeMask, Type, pCommandAllocator, pInitialState, IID_PPV_ARGS(&commandList));
+    if (D3DError(result, commandList, NULL, "Unable to create the Direct3D 12 command list")) {
+        return NULL;
+    }
+
+    d3d12_command_list *cmdList = malloct<d3d12_command_list>();
+    cmdList->p = commandList;
+    cmdList->signal = 0;
+
+    return cmdList;
+}
+
+static d3d12_command_list *new_compute_command_list(d3d12_device *device, d3d12_command_allocator *allocator) {
+    TRACELOG;
+    return new_command_list<HALIDE_D3D12_COMMAND_LIST_TYPE>(device, allocator);
+}
+
+static d3d12_copy_command_list *new_copy_command_list(d3d12_device *device, d3d12_command_allocator *allocator) {
+    TRACELOG;
+    return new_command_list<D3D12_COMMAND_LIST_TYPE_COPY>(device, allocator);
+}
+
+static d3d12_compute_pipeline_state *new_compute_pipeline_state_with_function(d3d12_device *device, d3d12_function *function) {
+    TRACELOG;
+    ID3D12PipelineState *pipelineState = NULL;
+    D3D12_COMPUTE_PIPELINE_STATE_DESC cpsd = { };
+        cpsd.pRootSignature = function->rootSignature;
```

Comment:
Fix indentation

Reply:
Actually, this indentation scheme is intentional (you'll see it in many places where I need to populate a D3D12 struct). At least for me, it improves readability, but I can re-indent it if that's better for the project. Should I?

Reply:
A way to preserve the way it reads without confusing people or editors might be an extra pair of braces:
```
FOO bar = {};
{
   bar.field1 = 34;
   bar.field2 = 54;
}
```

Reply:
I defer to @zvookin / other Halide devs on this, but it's nonstandard w.r.t. Halide code, so I'd unindent.

Reply:
I like Andrew's suggestion. If @zvookin has no objections, I'll follow Andrew's recommendation and apply the changes accordingly.